libs/utils/src/http/endpoint.rs

```diff
@@ -52,41 +52,31 @@ where
     /// Use as `|r| RequestSpan(my_handler).handle(r)` instead of `my_handler` as the request handler to get the span enabled.
     pub async fn handle(self, request: Request<Body>) -> Result<Response<B>, E> {
         let request_id = request.context::<RequestId>().unwrap_or_default().0;
+        let method = request.method();
+        let path = request.uri().path();
+        let request_span = info_span!("request", %method, %path, %request_id);
+
+        let log_quietly = method == Method::GET;
+        async move {
+            if log_quietly {
+                debug!("Handling request");
+            } else {
+                info!("Handling request");
+            }
+            let response = (self.0)(request).await?;
```

Comment:
There's some conversion which is probably still needed from `Result::Err` to body, otherwise we might not get the logging for handler returning an error cases?

Reply:
Oh, good catch, thank you.

I have a `Debug` bound on `E` and that might be a bit confusing, but best I could think of.
I'd need to use `:#?` formatter due to anyhow stacktraces possibility, but looks not bad:

```
2023-03-08T13:28:09.729221Z DEBUG parsed 3 headers
2023-03-08T13:28:09.729242Z DEBUG incoming body is empty
2023-03-08T13:28:09.729544Z DEBUG request{method=GET path=/v1/status request_id=2a045bd1-1002-4d24-a0e9-2b712a344f88}: Handling request
2023-03-08T13:28:09.729594Z  INFO request{method=GET path=/v1/status request_id=2a045bd1-1002-4d24-a0e9-2b712a344f88}: Status check called
2023-03-08T13:28:09.729634Z DEBUG request{method=GET path=/v1/status request_id=2a045bd1-1002-4d24-a0e9-2b712a344f88}: Status check debug message
2023-03-08T13:28:09.729969Z  WARN request{method=GET path=/v1/status request_id=2a045bd1-1002-4d24-a0e9-2b712a344f88}: Request failed with an error: InternalServerError(
    ";(((((((((",
)
2023-03-08T13:28:09.730014Z ERROR Error processing HTTP request: InternalServerError(;(((((((((

Stack backtrace:
   0: backtrace::backtrace::libunwind::trace
   .....
2023-03-08T13:28:09.786331Z DEBUG flushed 260 bytes
2023-03-08T13:28:09.786544Z DEBUG read eof
```

Trying to add another type error either via another struct generic or method generic did not work that well, so I abandoned the attempts.

Reply:
But the span will be lost on the error handler; it seems it should be exchanged for this?

Reply:
One extra concern: with carrying out the logs under new spans, we start to alter the library code that's used in safekeepers and proxy.

I'm not sure that fiddling with a global `err_handler` is good in these cases, so seems like we need to migrate these components either..

Reply:
Hmm... But if you don't log it under the same span, then it will look funky when eventually we have jaeger or whatever OTEL span visualizer, because it will look like a different span because the callsite identifies it.

I think this could be pageserver only functionality at first. Later on others can migrate if this works. I don't expect any of the middlewares can really work if we cannot really do this span attaching via a middleware. Perhaps the request-id header work is fine.

Reply:
I think that's starting to look very good. Also the case I mentioned is avoided because it's all in the same span.

I don't see the panic case from types, maybe add a comment about that to the `id tracker` if that's somewhere far away?

Reply:
I think you know better than me where `tracing_panic_hook` is 🙂 
Added a comment, mentioning it.

Reply:
Oki, looked at the `utils::http::error::handler`, understand the point here:

> only ApiErrors are allowed to be returned from internal handlers of the id tracker — otherwise we'll panic.

I think this is looking good. On top of that the expect in `utils::http::error::handler` will be logged within the span so all good.